src/normalize/normalize.rs

```diff
@@ -38,6 +43,52 @@ pub struct NormalizedAddrs<M> {
 pub type NormalizedUserAddrs = NormalizedAddrs<UserAddrMeta>;
 
 
+/// A type representing a build ID note.
+///
+/// In the ELF file, this header is typically followed by the variable sized
+/// build ID.
+#[repr(C)]
+struct BuildIdNote {
+    /// ELF note header.
+    header: Elf64_Nhdr,
+    /// NUL terminated string representing the name.
+    name: [c_char; 4],
+}
+
+// SAFETY: `BuildIdNote` is valid for any bit pattern.
+unsafe impl crate::util::Pod for BuildIdNote {}
+
+
+/// Attempt to read an ELF binary's build ID.
+// TODO: Currently look up is always performed based on section name, but there
+//       is also the possibility of iterating notes and checking checking
+//       Elf64_Nhdr.n_type for NT_GNU_BUILD_ID, specifically.
+fn read_build_id(path: &Path) -> Result<Option<String>> {
+    let file = File::open(path)?;
+    let parser = ElfParser::open_file(file)?;
+
+    // The build ID is contained in the `.note.gnu.build-id` section. See
+    // elf(5).
+    if let Ok(idx) = parser.find_section(".note.gnu.build-id") {
+        // SANITY: We just found the index so the section should always be
+        //         found.
+        let mut bytes = parser.section_data(idx).unwrap();
+        let _header = bytes.read_pod_ref::<BuildIdNote>();
+
+        // Every byte following the header maps to two hex characters.
+        let build_id = bytes
+            .iter()
+            .fold(String::with_capacity(bytes.len() * 2), |mut s, b| {
+                let () = s.push_str(&format!("{b:02x}"));
+                s
```

Comment:
do we need to do string conversion? why not return bytes? it's faster and we avoid questions of lowercase vs uppercase, etc. For simpler testing we can have a helper that converts it to string representation, of course.

Reply:
We don't strictly speaking need it. It felt more appropriate and was what the original design used, but we can postpone conversion. Will just use `Vec`.